### PR TITLE
Add partitioning

### PR DESCRIPTION
**Pull request by gburanov**

### CODE REVIEW

File: src/main/java/io/aiven/kafka/connect/gcs/HourlyTopicPartitionRecordGrouper.java [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+/**
+ * A {@link RecordGrouper} that groups records by topic and partition.
+ *
+ * <p>The class requires a filename template with {@code topic}, {@code partition},
+ * and {@code start_offset} variables declared.
+ *
+ * <p>The class supports limited and unlimited number of records in files.
+ */
+final class HourlyTopicPartitionRecordGrouper extends TopicPartitionRecordGrouper {
+    private static final List<String> EXPECTED_VARIABLE_LIST = Arrays.asList(
```
Comment (cedrics) [inline src/main/java/io/aiven/kafka/connect/gcs/HourlyTopicPartitionRecordGrouper.java:43]: Why not just construct the `HashSet` here?

File: src/main/java/io/aiven/kafka/connect/gcs/TopicPartitionRecordGrouper.java [hunk 2]
```diff
@@ -83,24 +84,27 @@ public void put(final SinkRecord record) {
...
             fileBuffers.computeIfAbsent(filename, ignored -> new ArrayList<>()).add(record);
         }
     }
 
-    private String renderFilename(final TopicPartition tp, final SinkRecord headRecord) {
+    protected boolean acceptsTemplateInstance(final Template filenameTemplate) {
+        return TopicPartitionRecordGrouper.acceptsTemplate(filenameTemplate);
+    }
+
+    protected Instance renderFilename(final TopicPartition tp, final SinkRecord headRecord) {
```
Comment (cedrics) [inline src/main/java/io/aiven/kafka/connect/gcs/TopicPartitionRecordGrouper.java]: Why did you move the render out of this function? Every caller stills calls render on it anyway.

If you don't want to perform the render inside this method, the name does not make sense anymore. Maybe sth. like `getBoundFilenameTemplate` (java names rock 😄 )

File: src/test/java/io/aiven/kafka/connect/gcs/GcsSinkTaskTest.java [hunk 3]
```diff
@@ -109,6 +110,35 @@ final void basic() {
             readSplittedAndDecodedLinesFromBlob("topic1-1-40", false, 0));
     }
 
+    @Test
+    final void basicNewTemplate() {
```
Comment (cedrics) [inline src/test/java/io/aiven/kafka/connect/gcs/GcsSinkTaskTest.java]: Maybe sth. like `withTimebasedGrouping` would be a better name here

File: src/main/java/io/aiven/kafka/connect/gcs/HourlyTopicPartitionRecordGrouper.java [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+        final LocalDateTime now = LocalDateTime.now();
+
+        return super.renderFilename(tp, headRecord)
+            .bindVariable(FilenameTemplateVariable.YEAR.name, () -> Integer.toString(now.getYear()))
+            .bindVariable(FilenameTemplateVariable.MONTH.name, () -> Integer.toString(now.getMonthValue()))
+            .bindVariable(FilenameTemplateVariable.DAY.name, () -> Integer.toString(now.getDayOfMonth()))
+            .bindVariable(FilenameTemplateVariable.HOUR.name, () -> Integer.toString(now.getHour()));
+    }
+
+        /**
```
Comment (cedrics) [inline src/main/java/io/aiven/kafka/connect/gcs/HourlyTopicPartitionRecordGrouper.java]: Indentation

**Review by cedrics — APPROVED** [review]
Logic LGTM, some minor complaints. Its been like 10 years since I looked at Java code 😬

File: src/main/java/io/aiven/kafka/connect/gcs/TopicPartitionRecordGrouper.java (hunk 2 above)
Comment (gburanov) [reply src/main/java/io/aiven/kafka/connect/gcs/TopicPartitionRecordGrouper.java]: I need to move render out because this method is overriden in the HourlyTopicPartitionRecordGrouper, where super is used to call render on this class, and then this render is extended.

File: src/main/java/io/aiven/kafka/connect/gcs/HourlyTopicPartitionRecordGrouper.java (hunk 1 above)
Comment (VOvchinnikov) [reply src/main/java/io/aiven/kafka/connect/gcs/HourlyTopicPartitionRecordGrouper.java:43]: AFAIR you can't instantiate HashSet as a static variable, but you definitely can do so in the constructor - and it has been declared as well.

Creating one every call is wasteful either way.

File: src/main/java/io/aiven/kafka/connect/gcs/HourlyTopicPartitionRecordGrouper.java [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+     * Checks if the template is acceptable for this grouper.
+     */
+    static boolean acceptsTemplate(final Template filenameTemplate) {
```
Comment (VOvchinnikov) [inline src/main/java/io/aiven/kafka/connect/gcs/HourlyTopicPartitionRecordGrouper.java:83]: static stuff should be before instance stuff pls

File: src/main/java/io/aiven/kafka/connect/gcs/TopicPartitionRecordGrouper.java (hunk 2 above)
Comment (VOvchinnikov) [reply src/main/java/io/aiven/kafka/connect/gcs/TopicPartitionRecordGrouper.java]: He did it for the extension in the class `HourlyTopicPartitionRecordGrouper`: there there is `return super.renderFilename(tp, headRecord)` call, result of which is later extended by hourly params.

However, this is a pretty poor way of extension, that harms readability of the code. 

Instead of removing the `.render()` call from the `renderFilename` method - which makes its name wrong - please, take the pre-`.render()` lines and extract them into a different method, like `bindFilenameTemplate` or something, close to what @cedrics wrote and then keep `.render` in the `renderFilename` method, instead of having it in every single caller.

File: src/test/java/io/aiven/kafka/connect/gcs/config/GcsSinkConfigTest.java [hunk 6]
```diff
@@ -178,10 +181,7 @@ void unsupportedOutputField() {
         final Throwable t = assertThrows(
             ConfigException.class, () -> new GcsSinkConfig(properties)
         );
-        assertEquals("Invalid value [key, value, offset, timestamp, unsupported] "
-                + "for configuration format.output.fields: "
-                + "supported values are: 'key', 'value', 'offset', 'timestamp'",
-            t.getMessage());
+        assertThat(t.getMessage(), StringContains.containsString("supported values are:"));
```
Comment (VOvchinnikov) [inline src/test/java/io/aiven/kafka/connect/gcs/config/GcsSinkConfigTest.java:184]: ahem, these changes are basically destroying the entire test.

Could they maybe be done in a little bit more subtle way? :) 

I admit, that original test is...questionable, but its destruction assures it won't be accepted in upstream, most likely.

**Review by VOvchinnikov — CHANGES_REQUESTED** [review]

File: src/main/java/io/aiven/kafka/connect/gcs/TopicPartitionRecordGrouper.java (hunk 2 above)
Comment (gburanov) [reply src/main/java/io/aiven/kafka/connect/gcs/TopicPartitionRecordGrouper.java]: ok maybe it is better in the terms of naming, I renamed it

File: src/test/java/io/aiven/kafka/connect/gcs/GcsSinkTaskTest.java (hunk 3 above)
Comment (gburanov) [reply src/test/java/io/aiven/kafka/connect/gcs/GcsSinkTaskTest.java]: 👍

File: src/main/java/io/aiven/kafka/connect/gcs/HourlyTopicPartitionRecordGrouper.java (hunk 4 above)
Comment (gburanov) [reply src/main/java/io/aiven/kafka/connect/gcs/HourlyTopicPartitionRecordGrouper.java]: 👍

File: src/main/java/io/aiven/kafka/connect/gcs/HourlyTopicPartitionRecordGrouper.java (hunk 5 above)
Comment (gburanov) [reply src/main/java/io/aiven/kafka/connect/gcs/HourlyTopicPartitionRecordGrouper.java:83]: this was not done by me, it is global pattern in all classes

File: src/test/java/io/aiven/kafka/connect/gcs/config/GcsSinkConfigTest.java (hunk 6 above)
Comment (gburanov) [reply src/test/java/io/aiven/kafka/connect/gcs/config/GcsSinkConfigTest.java:184]: I will try to make pr to upstream like that - lets see

**Comment by gburanov** [comment]
we do not use this solution anymore so closing it
